filters/auth/tokenintrospection.go

```diff
@@ -0,0 +1,342 @@
+// to validate authorization for requests. Current implementation uses
+// Bearer tokens to authorize requests and checks that the token
+// contains at least one key value pair provided.
+//
+// This is implementing RFC 7662 compliant implementation. It uses
+// POST requests to call introspection_endpoint to get the information
+// of the token validity.
+//
+// It uses /.well-known/openid-configuration path to the passed
+// oauthIssuerURL to find introspection_endpoint as defined in draft
+// https://tools.ietf.org/html/draft-ietf-oauth-discovery-06, if
+// oauthIntrospectionURL is a non empty string, it will set
+// IntrospectionEndpoint to the given oauthIntrospectionURL.
+func NewOAuthTokenintrospectionAllKV(cfg *OpenIDConfig, timeout time.Duration) filters.Spec {
+	return newOAuthTokenintrospectionFilter(checkOAuthTokenintrospectionAllKV, cfg, timeout)
+}
+
+func NewOAuthTokenintrospectionAnyClaims(cfg *OpenIDConfig, timeout time.Duration) filters.Spec {
+	return newOAuthTokenintrospectionFilter(checkOAuthTokenintrospectionAnyClaims, cfg, timeout)
+}
+
+func NewOAuthTokenintrospectionAllClaims(cfg *OpenIDConfig, timeout time.Duration) filters.Spec {
+	return newOAuthTokenintrospectionFilter(checkOAuthTokenintrospectionAllClaims, cfg, timeout)
+}
+
+func newOAuthTokenintrospectionFilter(typ roleCheckType, cfg *OpenIDConfig, timeout time.Duration) filters.Spec {
+	return &tokenIntrospectionSpec{
+		typ:              typ,
+		issuerURL:        cfg.Issuer,
+		introspectionURL: cfg.IntrospectionEndpoint,
+		config:           cfg,
+		timeout:          timeout,
+	}
+}
+
+func GetOpenIDConfig(issuerURL string) (*OpenIDConfig, error) {
+	u, err := url.Parse(issuerURL + TokenIntrospectionConfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	var cfg OpenIDConfig
+	err = jsonGet(u, "", &cfg, http.DefaultClient)
+	return &cfg, err
+}
+
+func (s *tokenIntrospectionSpec) Name() string {
+	switch s.typ {
+	case checkOAuthTokenintrospectionAnyClaims:
+		return OAuthTokenintrospectionAnyClaimsName
+	case checkOAuthTokenintrospectionAllClaims:
+		return OAuthTokenintrospectionAllClaimsName
+	case checkOAuthTokenintrospectionAnyKV:
+		return OAuthTokenintrospectionAnyKVName
+	case checkOAuthTokenintrospectionAllKV:
+		return OAuthTokenintrospectionAllKVName
+	}
+	return AuthUnknown
+}
+
+func (s *tokenIntrospectionSpec) CreateFilter(args []interface{}) (filters.Filter, error) {
+	sargs, err := getStrings(args)
+	if err != nil {
+		return nil, err
+	}
+	if len(sargs) == 0 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	ac, err := newAuthClient(s.introspectionURL, s.timeout)
+	if err != nil {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	f := &tokenintrospectFilter{
+		typ:        s.typ,
+		authClient: ac,
+		kv:         make(map[string]string),
+	}
+	switch f.typ {
+	case checkOAuthTokenintrospectionAllClaims:
+		fallthrough
+	case checkOAuthTokenintrospectionAnyClaims:
+		f.claims = sargs[:]
```

Comment:
wondering why this sargs[:] is required here, since we don't modify the sargs slice. Is there a reason?